### PR TITLE
Support for Yarn workspaces

### LINKED ISSUE #57: yarn workspaces
i've discovered [yarn workspaces](https://yarnpkg.com/blog/2017/08/02/introducing-workspaces/) in the project https://gitlab.coko.foundation/editoria/editoria/tree/master and i assume this is currently not handled by `yarn2nix`.

the problem is this:
* `editoria/package.json` (with `editoria/yarn.lock`) can be easily transformed into a node_package created by `yarn2nix`
* `editoria/packages/editoria-app/package.json` is a subdirectory/subproject which also uses the `yarn.lock` from the top-level but then generates a different `node_modules` output as it is based on a different `package.json`

i'm not sure how to cover this use-case with `yarn2nix` but hopefully only requires a little patch for `yarn2nix`. ideas?

### PR DESCRIPTION
**Pull request by nightkr**
This adds a new argument `workspaceDependencies` to `mkYarnPackage`, the contents  of which are added to a [Yarn Workspace](https://yarnpkg.com/lang/en/docs/workspaces/) before `yarn install`. Example:

```nix
  rec {
    package-one = yarn2nix.mkYarnPackage {
      src = ./workspace/package-one;
      yarnLock = ./workspace/yarn.lock;
      workspaceDependencies = { inherit package-two; };
    };
    package-two = yarn2nix.mkYarnPackage {
      src = ./workspace/package-two;
      yarnLock = ./workspace/yarn.lock;
    };
  }.package-one

```

# Notes:

* Workspace dependencies must be manually added to `workspaceDependencies`
  * However, transitive dependencies are supported (for a relationship `a -> b -> c`, `a` does not need to explicitly depend on `c`)
* Local dependency build steps are not supported
  * This is because local dependencies are removed at the end of `mkYarnModules` and added back in `mkYarnPackage`, to avoid re-running yarn after source-only changes

This (sort of) fixes #57.

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -45,12 +45,14 @@ in rec {
 
   mkYarnModules = {
     name,
+    pname,
```
Comment (zimbatm) [inline default.nix]: in that case let's remove the name argument and also pass the version along the `pname`

File: default.nix [hunk 2]
```diff
@@ -67,11 +69,26 @@ in rec {
...
+          " "
+          (x: "node_modules/${x}")
+          ([pname] ++ (lib.mapAttrsToList (name: x: x.pname) workspaceDependencies))}";
+in
     stdenv.mkDerivation {
-      inherit name preBuild;
+      inherit name preBuild workspaceJSON;
       phases = ["configurePhase" "buildPhase"];
       buildInputs = [ yarn nodejs ] ++ extraBuildInputs;
+      passAsFile = [ "workspaceJSON" ];
```
Comment (zimbatm) [inline default.nix]: This is an obscure feature of nix. Do you mind using `pkgs.writeText` to create a derivation out of `workspaceJSON` and pass this instead?

File: default.nix [hunk 3]
```diff
@@ -81,16 +98,21 @@ in rec {
       buildPhase = ''
         runHook preBuild
 
-        cp ${packageJSON} ./package.json
+        mkdir ${pname}
+        cp ${packageJSON} ./${pname}/package.json
+        cp $workspaceJSONPath ./package.json
         cp ${yarnLock} ./yarn.lock
-        chmod +w ./yarn.lock
+        chmod +w ./yarn.lock # ./package.json
```
Comment (zimbatm) [inline default.nix]: leftover comment

**Review by zimbatm — COMMENTED** [review]
Sounds good. It adds a lot of complexity but it seems like it's unavoidable. Thanks for doing all the research and implementing a solution for the workspaces scenario.

File: default.nix (hunk 2 above)
Comment (nightkr) [reply default.nix]: Sure.

File: default.nix (hunk 3 above)
Comment (nightkr) [reply default.nix]: Whoops, fixing.

**Comment by nightkr** [comment]
Done @zimbatm

**Comment by zimbatm** [comment]
thanks!

**Pull request merged**